### PR TITLE
Use s-parameter to build lumped components.

### PR DESCRIPTION
**Pull request by Asachoo**
Closes https://github.com/scikit-rf/scikit-rf/issues/1205

### CODE REVIEW

**Comment by Asachoo** [comment]
To validate the `S-Parameter` of lumped components, I have added three extra unit tests to assess mismatch scenarios.

**Comment by Ttl** [comment]
Does this handle complex port impedance correctly? The default `sdef` is `power` which usually causes issues with complex port impedance. It's often required to generate S-parameters with `traveling` definition and the uses `s2s` to change it to the default `power` definition.

It would be good to test against the current Y-parameter implementation. Test including complex port impedance would be also welcome.

**Review by mhuser — APPROVED** [review]
Looks good to me.

**Comment by Asachoo** [comment]
> Does this handle complex port impedance correctly? The default `sdef` is `power` which usually causes issues with complex port impedance. It's often required to generate S-parameters with `traveling` definition and the uses `s2s` to change it to the default `power` definition.
> 
> It would be good to test against the current Y-parameter implementation. Test including complex port impedance would be also welcome.

Thanks for your advice @Ttl ! Regrettably, the complex port impedance cases not yielded correct s-parameters in my local tests. I am considering addressing this issue by directly developing accurate S-parameters functions directly, but I worry that this approach may become more complex.

Alternatively, instead of directly constructing the mismatched Component, we could create a matched Component, such as a 50 Ohm, and then use `renormalize()` to accommodate the mismatched port impedance. Perhaps this approach could be more robust and make the code more comprehensible.

**Comment by mhuser** [comment]
Indeed `y2s` was applying different computations depending the `s_def` 🤔 
https://github.com/scikit-rf/scikit-rf/blob/3766c695dc46d8a9059e5ecad18714e50e2f0fb3/skrf/network.py#L7229-L7252

**Comment by Ttl** [comment]
See for example how `line` in `media` is calculated for reference. You can also calculate with fixed real `z0` first and then `renormalize` to correct `z0` and `sdef`. However, you need to make sure that `sdef` of returned network is correct, see again `line`.

**Comment by Asachoo** [comment]
Thanks ! @Ttl @mhuser 
I will reopen this PR once the `complex impedance` and `s_def` check are finalized.

**Comment by Asachoo** [comment]
With the default setting (i.e., `s_def='power'`) for S-parameters now correctly handling complex port impedances.
For more test details, refer to the updated unit tests.

**Comment by Ttl** [comment]
You have modified the equations to be correct for `s_def=power`, but I don't see how this can give correct result for any other `s_def`. You should have `renormalize` to `s_def = kwargs.pop('s_def', S_DEF_DEFAULT)` if other `s_def` is requested, just like in `line`. `result` network should have `s_def='power'` like in `line` with `pop` from `kwargs` before it, to make it easier to see what definition is used.

**Comment by Ttl** [comment]
Please also add test against the old Y-parameter definition like in #1207. It could also test with different `s_def`.

**Comment by Asachoo** [comment]
> You have modified the equations to be correct for `s_def=power`, but I don't see how this can give correct result for any other `s_def`. You should have `renormalize` to `s_def = kwargs.pop('s_def', S_DEF_DEFAULT)` if other `s_def` is requested, just like in `line`. `result` network should have `s_def='power'` like in `line` with `pop` from `kwargs` before it, to make it easier to see what definition is used.

The default `s_def` configuration (`s_def='power'`) and the `renormalize()` in necessary have been implemented.

> Please also add test against the old Y-parameter definition like in #1207. It could also test with different `s_def`.

The unit tests have been expanded to include checks for `real/imag Port impedance` renormalization, `s_def` renormalization, and comparisons with `Y-parameter references` to ensure accuracy and consistency.

Unexpectedly, two unexpected issues have arisen:
1. Under the condition of `s_def = 'pseudo'`, the S-parameter of the inductor shows an numerical error compared to the Y-parameter defined reference (with the decimal precision dropping from 6 to 5, but I think this is acceptable.).

2. There are unexpected and confusing unit test failures that require further investigation (But I didn't find out how this PR caused failures).

**Comment by mhuser** [comment]
@Asachoo the unit test error with `skrf/io/tests/test_ts_spec.py` is strange and does not seem to be linked at all with `Media`.
Can you try to modify line 63 of this file with the following?
```Python
ex_5_6 = rf.Network(s=np.tile(s, [2, 1, 1]), z0=[50, 75, 0.01, 0.01], f=[5, 6], f_unit="GHz")
```

**Review by mhuser — APPROVED** [review]
The S-parameters are now built for power-waves and the various `s_def` are handled by `renormalize`.

**Comment by Asachoo** [comment]
> @Asachoo the unit test error with `skrf/io/tests/test_ts_spec.py` is strange and does not seem to be linked at all with `Media`. Can you try to modify line 63 of this file with the following?
> 
> ```python
> ex_5_6 = rf.Network(s=np.tile(s, [2, 1, 1]), z0=[50, 75, 0.01, 0.01], f=[5, 6], f_unit="GHz")
> ```

My local test shows that Network Name and Frequency should not cause `AssertionError`
```python3
(scikit-rf) ➜  scikit-rf git:(FixIssue2) python   
Python 3.12.7 (main, Oct 16 2024, 07:12:08) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import skrf as rf
>>> import numpy as np
>>> rf.__file__
'/Users/asachoo/Git/scikit-rf/skrf/__init__.py'
>>> 
>>> s = np.random.rand(4,2,2)
>>> f1 = rf.Frequency(1,10,11,'ghz')
>>> f2 = rf.Frequency(1e9, 10e9,11,'hz')
>>> 
>>> # Different f unit
>>> ntwk1 = rf.Network(s=s, frequency=f1)
>>> ntwk2 = rf.Network(s=s, frequency=f2)
>>> ntwk1
2-Port Network: '',  1.0-10.0 GHz, 11 pts, z0=[50.+0.j 50.+0.j]
>>> ntwk2
2-Port Network: '',  1000000000.0-10000000000.0 Hz, 11 pts, z0=[50.+0.j 50.+0.j]
>>> ntwk1 == ntwk2
True
>>> 
>>> # Different Name
>>> ntwk1 = rf.Network(s=s, frequency=f1, name='withname')
>>> ntwk2 = rf.Network(s=s, frequency=f2)
>>> ntwk1
2-Port Network: 'withname',  1.0-10.0 GHz, 11 pts, z0=[50.+0.j 50.+0.j]
>>> ntwk2
2-Port Network: '',  1000000000.0-10000000000.0 Hz, 11 pts, z0=[50.+0.j 50.+0.j]
>>> ntwk1 == ntwk2
True
>>> 
```

**Comment by mhuser** [comment]
> My local test shows that Network Name and Frequency should not cause `AssertionError`

Indeed. Frequency is compared in Hz by `__eq__` operator. Pytest does not fail on my machine.
`platform win32 -- Python 3.9.16, pytest-7.1.2, pluggy-1.0.0` 🤔

**Comment by Asachoo** [comment]
Unable to pinpoint the cause of `test_ts_example_5_6()` failure 😔, all related fix attempts will be reverted.

**Comment by Ttl** [comment]
I'm not exactly sure what's the issue with the test failure either, but this passes [tests](https://github.com/Ttl/scikit-rf/actions/runs/11875719076/job/33093247093) on my branch: https://github.com/Ttl/scikit-rf/commit/b4c1aece4868a1ca87f56b94bae1584b5e9e1cb7.

**Comment by Asachoo** [comment]
Even the same commit (the latest 02f4f68) [passed actions](https://github.com/Asachoo/scikit-rf/actions/runs/11876295845) in my fork, but fails in the scikit repository, which is quite strange 🤔. Any suggestions @FranzForstmayr @jhillairet ?

**Comment by jhillairet** [comment]
> Even the same commit (the latest [02f4f68](https://github.com/scikit-rf/scikit-rf/commit/02f4f686c4a36704493681ac64fd5eba4a048891)) [passed actions](https://github.com/Asachoo/scikit-rf/actions/runs/11876295845) in my fork, but fails in the scikit repository, which is quite strange 🤔. Any suggestions @FranzForstmayr @jhillairet ?

Very strange indeed... I also passes locally on my windows machine and also on github with a copy of your branch on [my repo](https://github.com/jhillairet/scikit-rf/actions/runs/11878287492/job/33098786177)...

**Comment by Ttl** [comment]
It looks like it's an issue with Touchstone reading using uninitialized memory when loading "lower" or "upper" ts file. https://github.com/scikit-rf/scikit-rf/blob/master/skrf/io/touchstone.py#L536-L550. `np.empty` gives uninitialized array and it just happened to be zeroes previously. `np.nanmax` would then read the uninitialized array and error if entries weren't zero. For local testing replacing `L536`, `np.empty` with `np.ones` only breaks this test.

I'm not sure `np.nanmax` works correctly even when the data is initialized since max requires comparison of complex numbers and numpy complex comparison is lexical first real and then imaginary part. If input data has entries with negative real part it would fail here even if s was initialized to zero. It should be replaced with actual assignment of missing lower/upper triangular elements.

File: skrf/media/media.py [hunk 1]
```diff
@@ -643,14 +643,27 @@ def resistor(self, R: NumberLike, *args, **kwargs) -> Network:
         capacitor
         inductor
         """
-        result = self.match(nports=2, **kwargs)
-        y = np.zeros(shape=result.s.shape, dtype=complex)
+        s_def = kwargs.pop('s_def', S_DEF_DEFAULT)
+        result = self.match(nports=2, s_def=S_DEF_DEFAULT, **kwargs)
```
Comment (Ttl) [inline skrf/media/media.py:647]: I doubt we will change `S_DEF_DEFAULT` but to make sure it works if we do, this should be `s_def='power'`. Same in other classes.

**Review by Ttl — CHANGES_REQUESTED** [review]

File: skrf/media/media.py (hunk 1 above)
Comment (Asachoo) [reply skrf/media/media.py]: All S-Parameters that are only accurate under the definition of 'power' have been changed to explicitly `s_def='power'`.

Comment (Ttl) [reply skrf/media/media.py]: `kwargs.pop` should still have `S_DEF_DEFAULT` to make sure the correct `s_def` is returned if `S_DEF_DEFAULT` is not power.

Comment (Asachoo) [reply skrf/media/media.py]: Modified as suggested.

**Review by jhillairet — APPROVED** [review]

**Pull request merged**